dali/operators/audio/nonsilence_op.h

```diff
@@ -0,0 +1,208 @@
+  return LeadTrailThresh(make_cspan(signal_mms.data, signal_mms.num_elements()),
+                         dbc.db2signal(args.cutoff_db));
+}
+
+}  // namespace detail
+
+template<typename Backend>
+class NonsilenceOperator : public Operator<Backend> {
+ public:
+  ~NonsilenceOperator() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(NonsilenceOperator);
+
+ protected:
+  explicit NonsilenceOperator(const OpSpec &spec) :
+          Operator<Backend>(spec) {}
+
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+
+  void AcquireArgs(const OpSpec &spec, const workspace_t<Backend> &ws) {
+    this->GetPerSampleArgument(cutoff_db_, "cutoff_db", ws);
+    this->GetPerSampleArgument(reference_power_, "reference_power", ws);
+    window_length_ = spec.GetArgument<int>("window_length", &ws);
+    auto input_type = ws.template InputRef<Backend>(0).type().id();
+    // If input type is not floating point, there's no need for reset interval
+    reset_interval_ = IsFloatingPoint(input_type) ? spec.GetArgument<int>("reset_interval", &ws)
+                                                  : -1;
+
+    DALI_ENFORCE(reset_interval_ == -1 || reset_interval_ % window_length_ == 0,
+                 make_string("`reset_interval` shall be a multiple of `window_length`. "
+                             "Got: reset_interval: ", reset_interval_, " vs window_length: ",
+                             window_length_));
+  }
+
+
+  std::vector<float> cutoff_db_;
+  std::vector<float> reference_power_;
+  int window_length_ = -1;
+  int reset_interval_ = -1;
+
+  USE_OPERATOR_MEMBERS();
+};
+
+
+class NonsilenceOperatorCpu : public NonsilenceOperator<CPUBackend> {
+ public:
+  explicit NonsilenceOperatorCpu(const OpSpec &spec) :
+          NonsilenceOperator<CPUBackend>(spec) {
+    intermediate_buffers_.resize(num_threads_);
+  }
+
+
+  ~NonsilenceOperatorCpu() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(NonsilenceOperatorCpu);
+
+ protected:
+  bool SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
+                 const workspace_t<CPUBackend> &ws) override;
+
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+ private:
+  template<typename InputType>
+  void RunImplTyped(workspace_t<CPUBackend> &ws) {
+    const auto &input = ws.template InputRef<CPUBackend>(0);
+    auto &output_begin = ws.OutputRef<CPUBackend>(0);
+    auto &output_length = ws.OutputRef<CPUBackend>(1);
+    auto &tp = ws.GetThreadPool();
+
+    for (int sample_id = 0; sample_id < batch_size_; sample_id++) {
+      tp.DoWorkWithID(
+              [&, sample_id](int thread_id) {
+                  detail::Args<InputType> args;
+                  args.input = view<const InputType, 1>(input[sample_id]);
+                  args.cutoff_db = cutoff_db_[sample_id];
+                  args.reference_power = reference_power_[sample_id];
+                  args.reference_max = reference_power_[sample_id] == 0;
```

Comment:
I think we should rather check `HasArgument("reference_power")` at construction, to determine whether the maximum should be used or not, and `0.` should actually be an invalid argument (you can't divide by 0)

Reply:
Done